### PR TITLE
examples: FetchBlock: enhance this example to be able to connect dns discovered peers so that you don't need to install bitcoind locally

### PR DESCRIPTION
**Pull request by Bo-Ye**
In the examples, FetchBlock example only can retrieve a block from localhost, which is inconvenient and especially confusing for the newbies because localhost node needs to install bitcoind, etc.

This commit provides extra feature to download blocks from DNS seeds generating peers, by using --localhost flag on the command line.

Please review and leave comments, thanks.

### CODE REVIEW

**Comment by schildbach** [comment]
Thanks for your contribution! Generally I'm happy about this enhancement.

I'm not so happy about the introduced dependency to commons-cli. We already use [jopt-simple](https://pholser.github.io/jopt-simple/) for command line parsing, is there any reason to switch?

**Comment by Bo-Ye** [comment]
@schildbach Sorry for late reply, thanks for the comments, I didn't realise jopt-simple is available in the first place, I've corrected my PR as per your suggestion, please review again, many thanks.

The PR is rebased and squashed.

File: examples/src/main/java/org/bitcoinj/examples/FetchBlock.java [hunk 1]
```diff
@@ -17,34 +17,71 @@
...
 import org.bitcoinj.store.MemoryBlockStore;
 import org.bitcoinj.utils.BriefLogFormatter;
 
 import java.net.InetAddress;
+import java.util.List;
 import java.util.concurrent.Future;
 
 /**
- * Downloads the block given a block hash from the localhost node and prints it out.
+ * Downloads the block given a block hash from the remote/localhost node and prints it out.
```
Comment (schildbach) [inline examples/src/main/java/org/bitcoinj/examples/FetchBlock.java]: Paragraphs are missing closing tags. Also, the first paragraph needs markup too.

File: examples/src/main/java/org/bitcoinj/examples/FetchBlock.java [hunk 2]
```diff
@@ -17,34 +17,71 @@
...
+ * <p>
+ * Downloading from localhost node needs to install bitcoin https://bitcoin.org/en/download and run command: bitcoind -testnet -daemon
+ * After bitcoind up and running, use command: org.bitcoinj.examples.FetchBlock --localhost <blockHash>
```
Comment (schildbach) [inline examples/src/main/java/org/bitcoinj/examples/FetchBlock.java]: I suspect `<blockHash>` needs to be escaped.

File: examples/src/main/java/org/bitcoinj/examples/FetchBlock.java [hunk 3]
```diff
@@ -17,34 +17,71 @@
...
+            parser.accepts("localhost", "Connect to the localhost node");
+            parser.accepts("help", "Displays program options");
+            opts = parser.parse(args);
+            if (opts.has("help")) {
+                System.out.println("usage: org.bitcoinj.examples.FetchBlock [--localhost] <blockHash>");
+                parser.printHelpOn(System.out);
+                return;
+            }
+            nonOpts = opts.nonOptionArguments();
+            if(nonOpts.size() != 1){
```
Comment (schildbach) [inline examples/src/main/java/org/bitcoinj/examples/FetchBlock.java]: Space after `if`.

File: examples/src/main/java/org/bitcoinj/examples/FetchBlock.java [hunk 4]
```diff
@@ -17,34 +17,71 @@
...
+                throw new IllegalArgumentException("Incorrect number of block hash, please provide only one block hash.");
+            }
+        }catch (OptionException | IllegalArgumentException e) {
```
Comment (schildbach) [inline examples/src/main/java/org/bitcoinj/examples/FetchBlock.java]: Space before `catch`.

**Comment by schildbach** [comment]
I'm currently not sure why the test has failed, but it can't be caused by your changes.

**Comment by Bo-Ye** [comment]
Thanks for the comments, I've corrected them in the new commit which is rebased and squashed.

But this time the CI build failed again, I am not able to reproduce in my local, and it seems the failed tests are random, the failure this time is not same as failure last time.

Let me know any discoveries for the failing CI build, it's weird though. Thanks.

File: examples/src/main/java/org/bitcoinj/examples/FetchBlock.java (hunk 1 above)
Comment (Bo-Ye) [inline examples/src/main/java/org/bitcoinj/examples/FetchBlock.java]: good point, corrected.

File: examples/src/main/java/org/bitcoinj/examples/FetchBlock.java (hunk 2 above)
Comment (Bo-Ye) [inline examples/src/main/java/org/bitcoinj/examples/FetchBlock.java]: good point, corrected.

File: examples/src/main/java/org/bitcoinj/examples/FetchBlock.java (hunk 3 above)
Comment (Bo-Ye) [inline examples/src/main/java/org/bitcoinj/examples/FetchBlock.java]: good point, corrected.

File: examples/src/main/java/org/bitcoinj/examples/FetchBlock.java (hunk 4 above)
Comment (Bo-Ye) [inline examples/src/main/java/org/bitcoinj/examples/FetchBlock.java]: good point, corrected.

**Comment by Bo-Ye** [comment]
Finally, I got a green build after rebasing my pull request!

**Comment by schildbach** [comment]
Your test failures will likely be fixed with this PR: #1412. I'll try to merge it soon if no one objects.

I'll review your PR shortly.

**Comment by schildbach** [comment]
Merged.

**Comment by Bo-Ye** [comment]
cook, thanks.

**Pull request closed without merging**